src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -1349,9 +1349,20 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
           lambdaTarget.name.toString,
           methodBTypeFromSymbol(lambdaTarget).descriptor,
           /* itf = */ isInterface)
-      val receiver = if (isStaticMethod) Nil else lambdaTarget.owner :: Nil
-      val (capturedParams, lambdaParams) = lambdaTarget.paramss.head.splitAt(lambdaTarget.paramss.head.length - arity)
-      val invokedType = asm.Type.getMethodDescriptor(asmType(functionalInterface), (receiver ::: capturedParams).map(sym => typeToBType(sym.info).toASMType): _*)
+      val numCapturedParams = lambdaTarget.paramss.head.length - arity
+      val (capturedParams, lambdaParams) = lambdaTarget.paramss.head.splitAt(numCapturedParams)
+
+      val invokedTypeArgs: Array[asm.Type] = if (isStaticMethod) {
+        val result = new Array[asm.Type](numCapturedParams)
+        capturedParams.iterator.map(sym => typeToBType(sym.info).toASMType).copyToArray(result)
+        result
+      } else {
+        val result = new Array[asm.Type](numCapturedParams)
+        result(0) = typeToBType(lambdaTarget.owner.info).toASMType
+        capturedParams.iterator.map(sym => typeToBType(sym.info).toASMType).copyToArray(result, 1)
```

Comment:
Good catch. Turns out the second branch `(!isStatic`) is never taken, as we chose to make all lambda implementation methods static, which meant we could leave them public without having to mangle the names. But I've pushed a fix in case this becomes live code in the future.